### PR TITLE
added CMake options to provide external zlib, GRPC, Eigen

### PR DESCRIPTION
**Pull request by Slonegg**
Here are changes necessary to build tensorflow with different version of GRPC, Protobuf, Eigen, zlib, etc. It is not possible to compile project using two different versions of protobuf for example. To deal with that I've chosen approach similar to one in GRPC library. Copy of https://github.com/tensorflow/tensorflow/pull/14463#issuecomment-343576192

### CODE REVIEW

**Review by mrry — APPROVED** [review]
Looks good to me! Thanks for resubmitting.

**Comment by gunan** [comment]
I see this error in the build:
```
    40>CustomBuild:
         fatal: reference is not a tree: b04e5cba356212e4e8c66c61bbe0c3a20537c5b9
         CMake Error at T:/src/github/tensorflow/cmake_build/protobuf/tmp/protobuf-gitclone.cmake:65 (message):
           Failed to checkout tag: 'b04e5cba356212e4e8c66c61bbe0c3a20537c5b9'
```
Rerunning, but do you think there is a typo in the edited files?

File: tensorflow/contrib/cmake/external/protobuf.cmake [hunk 1]
```diff
@@ -14,38 +14,46 @@
 # ==============================================================================
 include (ExternalProject)
 
-set(PROTOBUF_INCLUDE_DIRS ${CMAKE_CURRENT_BINARY_DIR}/protobuf/src/protobuf/src)
-set(PROTOBUF_URL https://github.com/google/protobuf.git)
-set(PROTOBUF_TAG b04e5cba356212e4e8c66c61bbe0c3a20537c5b9)
+if (tensorflow_PROTOBUF_PROVIDER STREQUAL module AND tensorflow_GRPC_PROVIDER STREQUAL module)
+  set(PROTOBUF_INCLUDE_DIRS ${CMAKE_CURRENT_BINARY_DIR}/protobuf/src/protobuf/src)
+  set(PROTOBUF_URL https://github.com/mrry/protobuf.git)  # Includes MSVC fix.
```
Comment (mrry) [inline tensorflow/contrib/cmake/external/protobuf.cmake]: Minor tweak here: replace `mrry` with `google` to fix the build error.

**Review by mrry — CHANGES_REQUESTED** [review]
I think I found the source of the protobuf breakage...

File: tensorflow/contrib/cmake/external/protobuf.cmake (hunk 1 above)
Comment (Slonegg) [reply tensorflow/contrib/cmake/external/protobuf.cmake]: Updated, hope it helps.

**Comment by drpngx** [comment]
@Slonegg there is an error:

```
   168>PrepareForBuild:
         Creating directory "x64\Release\png_create_destination_dir\".
         Creating directory "x64\Release\png_create_destination_dir\png_crea.526FE762.tlog\".
       InitializeBuildStatus:
         Creating "x64\Release\png_create_destination_dir\png_crea.526FE762.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
       CustomBuild:
         Building Custom Rule T:/src/github/tensorflow/tensorflow/contrib/cmake/CMakeLists.txt
         CMake does not need to re-run because T:\src\github\tensorflow\cmake_build\CMakeFiles\generate.stamp is up-to-date.
   168>CUSTOMBUILD : CMake error : cmake version 3.7.2 [T:\src\github\tensorflow\cmake_build\png_create_destination_dir.vcxproj]
         Usage: C:\Program Files\cmake\bin\cmake.exe -E <command> [arguments...]
         Available commands: 
           capabilities              - Report capabilities built into cmake in JSON format
           chdir dir cmd [args...]   - run command in a given directory
           compare_files file1 file2 - check if file1 is same as file2
           copy <file>... destination  - copy files to destination (either file or directory)
           copy_directory <dir>... destination   - copy content of <dir>... directories to 'destination' directory
           copy_if_different <file>... destination  - copy files if it has changed
           echo [<string>...]        - displays arguments as text
           echo_append [<string>...] - displays arguments as text but no new line
           env [--unset=NAME]... [NAME=VALUE]... COMMAND [ARG]...
                                     - run command in a modified environment
           environment               - display the current environment
           make_directory <dir>...   - create parent and <dir> directories
           md5sum <file>...          - create MD5 checksum of files
           remove [-f] <file>...     - remove the file(s), use -f to force it
           remove_directory dir      - remove a directory and its contents
           rename oldname newname    - rename a file or directory (on one volume)
           server                    - start cmake in server mode
           sleep <number>...         - sleep for given number of seconds
           tar [cxt][vf][zjJ] file.tar [file/dir1 file/dir2 ...]
                                     - create or extract a tar or zip archive
           time command [args...]    - run command and return elapsed time
           touch file                - touch a file.
           touch_nocreate file       - touch a file but do not create it.
         Available on Windows only:
           delete_regv key           - delete registry value
           env_vs8_wince sdkname     - displays a batch file which sets the environment for the provided Windows CE SDK installed in VS2005
           env_vs9_wince sdkname     - displays a batch file which sets the environment for the provided Windows CE SDK installed in VS2008
           write_regv key value      - write registry value
         
   168>Done Building Project "T:\src\github\tensorflow\cmake_build\png_create_destination_dir.vcxproj" (default targets) -- FAILED.
```

File: tensorflow/contrib/cmake/external/png.cmake [hunk 2]
```diff
@@ -14,49 +14,60 @@
...
+			-DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=${tensorflow_ENABLE_POSITION_INDEPENDENT_CODE}
+			-DCMAKE_BUILD_TYPE:STRING=Release
+			-DCMAKE_VERBOSE_MAKEFILE:BOOL=OFF
+			-DCMAKE_INSTALL_PREFIX:STRING=${png_INSTALL}
+			${PNG_ADDITIONAL_CACHE_ARGS}
+	)
+
+	## put png includes in the directory where they are expected
+	add_custom_target(png_create_destination_dir
+		COMMAND ${CMAKE_COMMAND} -E make_directory ${png_INCLUDE_DIR}
```
Comment (mrry) [inline tensorflow/contrib/cmake/external/png.cmake]: I think `${png_INCLUDE_DIR}` here needs to be `${png_INCLUDE_DIRS}`.

**Review by mrry — APPROVED** [review]

**Comment by rmlarsen** [comment]
@drpngx @mrry Thanks, guys.

**Comment by drpngx** [comment]
@mrry looks like the same error...

**Comment by mrry** [comment]
Could be due to case sensitivity? The mkdir command has `${png_INCLUDE_DIRS}` whereas it's defined as `${PNG_INCLUDE_DIRS}`.

**Comment by jhseu** [comment]
Mind rebasing and fixing the errors? Thanks!

**Comment by martinwicke** [comment]
@Slonegg can you look into the test failures (likely case mismatch), and rebase?

Thanks!

**Comment by Slonegg** [comment]
Too many conflicts, sorry folks, I'm tired.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
Marking as abandoned. Feel free to reopen if you manage to make progress.

**Comment by ewilderj** [comment]
This is the sort of bug that we can advertise to the SIG Build list (tf-distribute as-was) as help wanted. I will send a message accordingly.

**Pull request closed without merging**